### ISSUE TITLE
Docker fails to start after v3.0.0 upgrade

### ISSUE THREAD

**Issue by tonyaware** [docker/for-win#9755]
- [X] I have tried with the latest version of my channel (Stable or Edge)
  - [X] I have uploaded Diagnostics
  - Diagnostics ID:
  506E8EEA-0D57-4A88-9EDC-E7C01916822A/20201210184243

### Expected behavior
docker starts
### Actual behavior
docker fails to start after ~3 min
### Information

  - Windows Version: windows 10 v1909 (OS Build 18363.1256)
  - Docker Desktop Version: v3.0.0
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

### Steps to reproduce the behavior

  1. upgrade from Version: 2.5.0.1 (49550) to Version: 3.0.0 (50684)
  2. attempt to start docker
3. even after reboot cannot start docker

---

EDIT by @tiborvass:

See 🚧 Workaround 🚧 : https://github.com/docker/for-win/issues/9755#issuecomment-748539590

**Comment by vdobrikov** [comment]
I faced with the same issue, but after `Troubleshoot -> Reset to factory defaults` it started.
MacBook 13" i5 BigSur

**Comment by chowchowww** [comment]
I faced with the same issue, my old docker after failed to install 3.0 version it is restarting for 30 mins

**Comment by SlavidDennis** [comment]
Similar issue here:
- Tried to update
- update asked to kill System
- Update failed
- Docker won't start

**Comment by chowchowww** [comment]
I solved the problem by:
- Uninstall completely Docker about twice times ( first time was failed )
- Reinstall Docker v3.0
But new problem here: 
- I tried both WSL2 & Hyper-V
+ WSL2 is faster but when i tried to reduce the ram & processors, it didnt work ( 100% CPU ~90% Ram ). I write a .wslconfig but still got 100% CPU
+ Back to Hyper-V: still slow as usual :(

**Comment by SlavidDennis** [comment]
Uninstall solved it for me too. Also required to attempts to uninstall, as the installer tried to kill System again :)
Didn't try WSL2

**Comment by silasary** [comment]
Factory Reset fixed it for me

**Comment by adnoh** [comment]
loosing all data is not really a fix, is it? What raises my attention, the VM in Hyper-V is created (after docker desktop tries to start) WITHOUT any network adapters!? Bug or feature?

**Comment by kennethnipper** [comment]
Same issue here.  I have important containers and images that I don't want to lose, so that's hopefully not required.  Any other suggestions?

**Comment by tonyaware** [comment]
I (Bug Creator), too had to uninstalled v3 and reinstall v3.  In doing so I lost my containers and a full day's of work to rebuild my local dev environment.  When I reinstalled I did not activate WSL2 integration.  FWIW, it was disabled in v2.5 prior the upgrade to v3.

**Comment by adnoh** [comment]
for me it seems like it is fixed without loosing any data, this is what I did:

- open docker desktop
- open hyper v and manually start the vm
- create a new VM to mount the docker VHD and access the filesystem of the mobyVM (SKIP THIS STEP)
- went out with the dogs and shout out loud my anger (THIS HELPED)
- Come back to see if the new VM is finished, to realize now my docker desktop says "running" and with a "docker ps" I started to celebrate an early christmas because everything was back running!
hopefully anyone has also luck with this. I guess it is some wired timing issue when the vm starts with the docker desktop. just leave it running, restart the vm manually in hyper v and drink a couple of coffee....

**Comment by geoff604** [comment]
I had the issue when I upgraded my local Docker Desktop on Windows to version 3.0.0, Docker failed to start. The trick was to kill the Hyper V container for DockerDesktopVM using task manager following the instructions here: http://woshub.com/how-to-stop-a-hung-virtual-machine-on-hyper-v-2016/ and then restart Docker Desktop. 

It seems that if DockerDesktopVM is set to start up automatically when windows starts, it starts before Docker Desktop runs, and then Docker Desktop tries to stop the already running VM which hangs, and then Docker Desktop isn't able to start the VM again. 

So to prevent this from happening again, you can try changing "Automatic Start Action" in Hyper-V Manager settings for DockerDesktopVM to "None".

**Comment by snowbear2014** [comment]
I updated docker to 3.0.0 yesterday, after reboot my laptop today, docker cannot start. I remembered last time docker failed to start with timeout and I made some changes to Window security settings according to this post: https://github.com/docker/for-win/issues/4393 suggested by rezapci:
1, Open "Window Security"
2, Open "App & Browser control"
3, Click "Exploit protection settings" at the bottom
4, Switch to "Program settings" tab
5, Locate "C:\WINDOWS\System32\vmcompute.exe" in the list and expand it
6, Click "Edit"
7, Scroll down to "Code flow guard (CFG)" and uncheck "Override system settings"
8, Start vmcompute from powershell "net start vmcompute"
So I rolled back the changes (basically check "Override system settings") again and restarted docker, it seems fixed the issue.

Not sure if it applies to everyone. but hope it can help someone in the same situation as me.

**Comment by TheSench** [comment]
Doing a factory reset also fixed the issue for me.  One thing I noticed was that this setting got added to the Docker Engine JSON:
```json
"features": {
  "buildkit": true
}
```
It's possible that adding this may fix the issue for someone else without them having to do a factory reset.

**Comment by stephen-turner** [comment]
Thanks for the reports. We're looking at this, but there's not much we can do without diagnostics having been uploaded. Could people who have seen this error please upload your diagnostics and paste the diagnostics id here? Thanks.

**Comment by ldrag** [comment]
Also having problems to start docker after upgrading to 3.0.
The uploaded diagnostic id:
6CDE51DA-9497-4DA0-A94D-DCA90F098CC2/20201214141302

**Comment by gedoor** [comment]
96259DCD-5791-4846-8DFD-051C42D5E004/20201215010545

**Comment by timonltv** [comment]
5E10E3ED-3D8B-467C-A2EB-526AD4300D53/20201215075453

**Comment by Matts** [comment]
E33868D1-76BE-4418-AB6C-FBD56CDAD954/20201215093624

**Comment by stephen-turner** [comment]
An update on this ticket: we have narrowed it down to a change of behaviour in the startup sequence of the new 20.10 engine. We're now talking to the engine team to fix it.

**Comment by speti43** [comment]
Same here, factory reset is just a temporary fix till next windows restart, and i loose all my data... Please fix this asap. Thank you!

**Comment by kennethnipper** [comment]
Really need a "rollback" option for failed updates like this.

**Comment by BenHarris** [comment]
Identical failed to restart errors to @kennethnipper

Diagnostic ID: 503C1AC2-E5B3-477D-92E2-66784DC4D48C/20201217163346

@stephen-turner do you have an ETA on when a patch will be released? and is there an interim solution which doesn't require a factory reset?

**Comment by stephen-turner** [comment]
No, we don't have the root cause yet and have been unable to reproduce it ourselves. But we're working on it.

**Comment by BenHarris** [comment]
> No, we don't have the root cause yet and have been unable to reproduce it ourselves. But we're working on it.

Happy to help in any way I can. Mine is still failing and I haven't done a reset or reinstall yet.

**Comment by stephen-turner** [comment]
@BenHarris There might be.
@simonferquel Are there any ways people on this ticket can help?

**Comment by simonferquel** [comment]
Hmm @BenHarris, if you don't have sensitive information in your volumes / containers, could you please export your docker-desktop-data distro and send its content to me (simon.ferquel@docker.com)?
To export it, just run wsl.exe --export docker-desktop-data export-file.tar.

**Comment by shawinder** [comment]
It worked for me after upgrade for few days `until I restarted my windows` and then it was simply stopping and shutting down on every restart. Tried deleting Hyper-V container and then restarting docker. It was creating the VM but again stopping and shutting down. (Sorry forgot to copy the logs).
I had to uninstall/re-install to make it work. But now I have lost all data (lots of configuration settings) in my volumes 👎

**Comment by gedoor** [comment]
mcr.microsoft.com/mssql/server:2019-latest
this docker-data

**Comment by simonferquel** [comment]
I can reproduce, fantastic !, thank you !

**Comment by simonferquel** [comment]
Issue identified, fix inbound, thanks all ! https://github.com/moby/moby/pull/41817

**Comment by BenHarris** [comment]
Thank you so much for your hard work @simonferquel! Do you know long this is likely to take to propagate through? Is there anything we can do in the meantime?

**Comment by simonferquel** [comment]
Difficult to say with the holliday period, but we'll have a preview build probably early next week.

**Comment by kayest** [comment]
I have probably the most unique scenario of all. I can't open Docker for Windows or Doctor Desktop. I also can't uninstall or install Docker Desktop. I think my only option is to find out how to uninstall Docker Desktop manually and start fresh.

**Comment by tiborvass** [comment]
There is a fix incoming at moby/moby#41817 thanks to @simonferquel and @cpuguy83, in the meanwhile this is the workaround, I've tested myself.

# 🚧  WORKAROUND 🚧 

```
wsl.exe -d docker-desktop
$ nsenter -a -t $(pgrep lifecycle-server) find /var/lib/docker/containers -name config.v2.json -exec sed -i'' -E 's/"Running":true(,.*"Restarting":true)/"Running":false\1/' {} \;
```

**Comment by italolemos** [comment]
> I faced with the same issue, but after `Troubleshoot -> Reset to factory defaults` it started.
> MacBook 13" i5 BigSur

This solved my problem when I went to 3.0.3 version on mac

**Comment by kennethnipper** [comment]
Quick question.  Does this apply to docker-ce on linux as well?  I have a linux box and that update is still waitiing on me to hit update.

**Comment by shawinder** [comment]
> 
> 
> It worked for me after upgrade for few days `until I restarted my windows` and then it was simply stopping and shutting down on every restart. Tried deleting Hyper-V container and then restarting docker. It was creating the VM but again stopping and shutting down. (Sorry forgot to copy the logs but the main error was `cancellationtoken triggered before we finished reading from the stream`).
> I had to uninstall/re-install to make it work. But now I have lost all data (lots of configuration settings) in my volumes 👎
> Edition Windows 10 Pro
> Version 20H2
> Installed on ‎6/‎4/‎2020
> OS build 19042.685
> Experience Windows Feature Experience Pack 120.2212.551.0

Today `after restarting my  machine` docker failed again with the same error! I have uploaded the diagnosis report. `8C656AD0-D9E7-4422-9917-EFB9B22097BA/20201228175815`

**Comment by shawinder** [comment]
> 
> 
> There is a fix incoming at [moby/moby#41817](https://github.com/moby/moby/pull/41817) thanks to @simonferquel and @cpuguy83, in the meanwhile this is the workaround, I've tested myself.
> # 🚧 WORKAROUND 🚧
> 
> ```
> wsl.exe -d docker-desktop
> $ nsenter -a -t $(pgrep lifecycle-server) find /var/lib/docker/containers -name config.v2.json -exec sed -i'' -E 's/"Running":true(,.*"Restarting":true)/"Running":false\1/' {} \;
> ```
> 
> Then Quit Docker Desktop, and start it again. (Maybe Restart works too, I didn't try that).
> 
> Please note that this is due to a container that has a restart policy, and has been restarting probably for a while. Most likely they will continue to exit early so do check the logs of your containers for some application error that needs to be handled. If there's anything good coming out of this, is that it may help you identify an application error. Sorry again for this uncool upgrade experience :(

I tried it but getting a following error:
```
nsenter: failed to parse pid: 'find'
```

**Comment by fcFn** [comment]
@shawinder Had the same issue. Try quitting Docker Desktop, start it again and run the command immediately after that.

**Comment by shawinder** [comment]
Thanks @fcFn I actually complete uninstalled/re-installed docker and was getting errors consistently on every shutdown/restart of the PC. In the end I disabled WSL altogether in windows features and have only Hyper-V turned on. Now it looks more stable than before.

**Comment by learning** [comment]
I can't use docker for windows anymore after update to 3.0, uninstall, reinstall and reset are not works.

**Comment by stephen-turner** [comment]
We have made a new release (3.0.4) with the latest Docker engine, and we believe that fixes this bug.

**Comment by speti43** [comment]
Cofirmed, It works again after the update. Thank you!

**Comment by stephen-turner** [comment]
Great, thanks.

**Comment by SiddAjriY2Jaccount** [comment]
I have the following version of Docker:
```
Docker version 17.03.1-ce-rc1, build 3476dbf
```
And I have "Docker for Windows" installed (not Docker Desktop). Everytime I attempted to update to 3.0.0, it would just not do anything.

Now with the latest release, can I proceed with downloading and installing Docker Desktop 3.0.4 from Docker Hub or do I need to uninstall my existing "Docker for Windows" version?

**Comment by lachlankrautz** [comment]
I still for the lifecycle error after this update and then a windows restart. Still had to:

- uninstall docker
- delete all AppData dirs
- re install docker

To fix.

**Comment by iisisrael** [comment]
Continued to fail even after the 3.0.4 upgrade.  @TheSench's [solution above](https://github.com/docker/for-win/issues/9755#issuecomment-744066866) worked.  Had to [create the config file](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon#configure-docker-with-a-configuration-file) at C:\ProgramData\Docker\config\daemon.json, with only the contents of:

    {
        "features": {
            "buildkit": true
        }
    }

I then killed the Docker.Service process in Task Manager.  Started Docker Desktop, and it was quickly running.

**Comment by jahwood** [comment]
Nothing from above helps :(

Version: 3.1.0 (51484) 
Build: 19042
OS: Windows 10 Pro
Edition: Professional

"Docker starting ..." nothing happens, no errors

**Comment by rpasche** [comment]
> Nothing from above helps :(
> 
> Version: 3.1.0 (51484)
> Build: 19042
> OS: Windows 10 Pro
> Edition: Professional
> 
> "Docker starting ..." nothing happens, no errors

Yes...same issue on my machine.

**Comment by uberubert** [comment]
My 3.1, freshly installed on a freshly started computer, keeps crashing and becoming unresponsive.

It seems to work fine for a while, like 15-30 minutes, then the container stops responding. The container then can't be restarted or deleted. Even a fresh restart without any containers seem to produce this result, where docker itself seems to become unresponsive.

I can't get the docker service started any longer. Even the "Reset to defaults" generates the crash report window.
Currently stuck on trying to uninstall docker, which obviously also has crashed. Will try reverting to some older version again once I can get it uninstalled.

**Comment by cpuguy83** [comment]
Any logs anywhere for these issues?

**Comment by speti43** [comment]
I'll skip this build then. Thanks for the notification! Updating docker nowadays is a very brave venture... Or downright crazy.

**Comment by uberubert** [comment]
Update: I finally figured out my problem. While my problem should probably be its own thread, I'll post the solution here since I already posted about the problem here.

My issue: WSL 2 integration was enabled after updating, but no WSL 2 distro was installed. I used Hyper-V up until now, atleast that's what I believe. Docker settings were mostly unavailable due to being stuck in "restarting" or "shutting down".

I was confused because it seemed stable and working as some of my containers miraculously started and even worked fine for a while.  I finally realised what the problem was once I was able to access the settings and saw the "You don't have any WSL 2 distro" message. I also noticed some settings were changing after some of my restarts, like the subnet mask was suddenly empty.

My solution: Followed [this microsoft guide](https://docs.microsoft.com/en-us/windows/wsl/install-win10) for setting up WSL 2 from scratch, step by step, ending with adding ubuntu from the store. Made sure docker used the ubuntu for wsl integration. Containers started behaving a lot better.

Note that I'm also struggling with a tiny C: drive partition, from which I have to remember to move my docker-data manually (wasn't this a setting before?). If I forgot this in between a few reinstall-attempts, it may have further increased the amount of problems I were facing and may possibly even explain some of them.
